integration/spark/vendor/iceberg/src/main/java/io/openlineage/spark/agent/vendor/iceberg/metrics/CatalogMetricsReporterHolder.java

```diff
@@ -0,0 +1,150 @@
+        openLineageMetricsReporter = new OpenLineageMetricsReporter(existing);
+      } else {
+        log.debug("No existing metrics reporter found");
+        openLineageMetricsReporter = new OpenLineageMetricsReporter();
+      }
+
+      // set metrics reporter in context
+      catalogMetricsReporter.put(catalog.name(), openLineageMetricsReporter);
+
+      // set metrics reporter in catalog
+      metricsReporterField.set(catalog, openLineageMetricsReporter);
+      log.info("Injected metrics reporter into Iceberg catalog");
+    } catch (IllegalAccessException e) {
+      log.warn("Unable to inject metrics reporter", e);
+    }
+  }
+
+  /**
+   * Get the commit report for the given snapshot id. If the report is found, it is removed from the
+   * reporter.
+   *
+   * @param snapshotId snapshot id
+   */
+  public Optional<IcebergScanReportInputDatasetFacet> getScanReportFacet(long snapshotId) {
+    Optional<IcebergScanReportInputDatasetFacet> scanReportInputDatasetFacet =
+        catalogMetricsReporter.values().stream()
+            .flatMap(reporter -> reporter.getScanReportFacets().stream())
+            .filter(report -> report.getSnapshotId() == snapshotId)
+            .findFirst();
+
+    if (scanReportInputDatasetFacet.isPresent()) {
+      log.debug("Returning scan report facet for snapshot id: {}", snapshotId);
+    } else {
+      log.debug("No scan report facet found for snapshot id: {}", snapshotId);
+    }
+
+    return scanReportInputDatasetFacet;
+  }
+
+  /**
+   * Get the commit report for the given snapshot id. If the report is found, it is removed from the
+   * reporter.
+   *
+   * @param snapshotId snapshot id
+   */
+  public Optional<IcebergCommitReportOutputDatasetFacet> getCommitReportFacet(long snapshotId) {
+    Optional<IcebergCommitReportOutputDatasetFacet> commitReportOutputDatasetFacet =
+        catalogMetricsReporter.values().stream()
+            .flatMap(reporter -> reporter.getCommitReportFacets().stream())
+            .filter(facet -> facet.getSnapshotId() == snapshotId)
+            .findFirst();
+
+    if (log.isDebugEnabled()) {
+      if (catalogMetricsReporter.isEmpty()) {
+        log.debug("Catalog metrics reporter is empty");
+      }
+
+      if (commitReportOutputDatasetFacet.isPresent()) {
+        log.debug("Returning commit report facet for snapshot id: {}", snapshotId);
+      } else {
+        log.debug("No commit report facet found for snapshot id: {}", snapshotId);
+      }
+    }
+
+    return commitReportOutputDatasetFacet;
+  }
+
+  @VisibleForTesting
+  OpenLineageMetricsReporter getReporterFor(String catalogName) {
+    return catalogMetricsReporter.get(catalogName);
+  }
+
+  @VisibleForTesting
+  void clear() {
+    catalogMetricsReporter.clear();
+  }
+
+  private static class SingletonHolder {
+    public static final CatalogMetricsReporterHolder instance = new CatalogMetricsReporterHolder();
```

Comment:
Rather than doing singleton here, put it into `OpenLineageContext`

Reply:
Let me look into this again.

Reply:
What we need to do is: register OpenLineage metrics reporter within the Iceberg catalog while still having a handle to it. Singleton is the easiest handle to go. Whenever a dataset's version facet is added, we call custom facet builders which attach iceberg metrics facets based on the version. 

Including metrics reporter in `OpenLineageContext` is more complex, mainly bcz for each run (each run a new context is created), we need to get the catalog from the logical plan, get the metrics reporter attached to a catalog and link it with current context. I believe this is doable and better, but it's not a straight-forward code change.